shell/platform/linux/fl_engine.cc

```diff
@@ -8,19 +8,108 @@
 
 #include <gmodule.h>
 
+static constexpr int kMicrosecondsPerNanosecond = 1000;
+
+// Unique number associated with platform tasks
+static constexpr size_t kPlatformTaskRunnerIdentifier = 1;
+
 struct _FlEngine {
   GObject parent_instance;
 
+  // Thread the GLib main loop is running on
+  GThread* thread;
+
   FlDartProject* project;
   FlRenderer* renderer;
   FLUTTER_API_SYMBOL(FlutterEngine) engine;
+
+  // PlatformMessageHandler keyed by channel name
+  GHashTable* platform_message_handlers;
+
+  // Messages being processed
+  GPtrArray* message_handles;
 };
 
 G_DEFINE_QUARK(fl_engine_error_quark, fl_engine_error)
 
 G_DEFINE_TYPE(FlEngine, fl_engine, G_TYPE_OBJECT)
 
-// Callback from Flutter engine that are passed to the renderer
+// Subclass of GSource that integrates Flutter tasks into the GLib main loop
+typedef struct {
+  GSource parent;
+  FlEngine* self;
+  FlutterTask task;
+} FlutterSource;
+
+typedef struct {
+  FlEnginePlatformMessageCallback callback;
+  gpointer user_data;
+} PlatformMessageHandler;
+
+PlatformMessageHandler* platform_message_handler_new(
+    FlEnginePlatformMessageCallback callback,
+    gpointer user_data) {
+  PlatformMessageHandler* handler = static_cast<PlatformMessageHandler*>(
+      g_malloc0(sizeof(PlatformMessageHandler)));
+  handler->callback = callback;
+  handler->user_data = user_data;
+  return handler;
+}
+
+void platform_message_handler_free(gpointer data) {
+  PlatformMessageHandler* handler = static_cast<PlatformMessageHandler*>(data);
+  g_free(handler);
+}
+
+struct _FlEnginePlatformMessageHandle {
+  const FlutterPlatformMessage* message;
+  GBytes* data;
+};
+
+static FlEnginePlatformMessageHandle* message_handle_new(
+    const FlutterPlatformMessage* message) {
+  FlEnginePlatformMessageHandle* handle =
+      static_cast<FlEnginePlatformMessageHandle*>(
+          g_malloc0(sizeof(FlEnginePlatformMessageHandle)));
+  handle->message = message;
```

Comment:
It looks like you're keeping a copy of a struct full of dangling pointers here. Anything you want to put into a reference-counted object from `message` needs to be copied.

Reply:
It wasn't clear to me the ownership of `FlutterPlatformMessage` - I assumed that is is valid until `FlutterEngineSendPlatformMessageResponse` is called? Does the `uint8_t*` data also need to be copied?

Reply:
Answered in Discord, but to capture here: the lifetime of parameters is generally only guaranteed a during a call. We can't assume that the message is valid until the response is provided.

(In the case of the message, it's very common for that data to only be needed synchronously during handling of the message receipt, even if the response is async, so it would be inefficient for the engine to always store it.)